apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/ClientPoolTests.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Spanner.V1;
+using Google.Cloud.Spanner.V1.Logging;
+using Moq;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Google.Cloud.Spanner.Data.Tests
+{
+    public class ClientPoolTests
+    {
+        public ClientPoolTests(ITestOutputHelper outputHelper)
+        {
+            TestLogger.TestOutputHelper = outputHelper;
+            TestLogger.Install();
+            Logger.LogLevel = V1.Logging.LogLevel.Debug;
+        }
+        private static Mock<ISpannerClientFactory> SetupMockClientFactory(Mock<SpannerClient> mockClient)
+        {
+            var mockClientFactory = new Mock<ISpannerClientFactory>();
+            mockClientFactory.Setup(
+                    x => x.CreateClientAsync(It.IsAny<ServiceEndpoint>(), It.IsAny<ITokenAccess>()))
+                .ReturnsAsync(
+                    () =>
+                    {
+                        //return a unique instance each time.
+                        var result = mockClient.Object;
+                        mockClient = new Mock<SpannerClient>();
+                        return result;
+                    });
+            return mockClientFactory;
+        }
+
+        [Fact]
+        public async Task ClientsCreatedOnDemandAsync()
+        {
+            //SpannerClients should not be precreated, but be created lazily.
+            var firstReturnedClient = new Mock<SpannerClient>();
+            var mockClientFactory = SetupMockClientFactory(firstReturnedClient);
+
+            var testPool = new ClientPool(mockClientFactory.Object);
+            var clientAcquired = await testPool.AcquireClientAsync();
+            testPool.ReleaseClient(clientAcquired);
+
+            mockClientFactory.Verify(x => x.CreateClientAsync(It.IsAny<ServiceEndpoint>(), It.IsAny<ITokenAccess>()), Times.Exactly(1));
+            var s = new StringBuilder();
+            Assert.Equal(0, testPool.GetPoolInfo(s));
+            Logger.Instance.Info(s.ToString());
+        }
+
+        [Fact]
+        public async Task SequentialCreatesOnlyUseOneChannel()
+        {
+            //SpannerClients if created sequentially should only use a single channel.
+            //This is to faciliate good session pool hits (which is per channel).  If we always
+            //round robin'd clients, then we would get hard cache misses until we populated all
+            //of the caches per channel.
+            //Since channels are lazily created, this also saves resources for scenarios like rich clients
+            //that may only need a single session for multiple sequential workloads.
+            //Our default setting for # channels = 4 needs to work well for all scenarios.
```

Comment:
same client = same channel.  so by asserting that the same client is returned each time, you know that its resuing the same channel.  I'll be sure to copy this comment in the implementation, but this context is important to know if/when the test fails.